### DISCUSSION TITLE
How to pass storage buffer array to a function

### DISCUSSION THREAD

**Discussion by skbizport** [gpuweb/gpuweb#4309] [Q&A] [answered]
Hi,

How would one go about passing array to a function? Here is a use case

1. Multiple dynamic storage arrays passed as input to a compute shader, say array1, array2
2. A  shared library function say utility() takes an array and performs some operations and returns a value
3. Multiple compute functions say compute1(), compute2() are used. Each compute function does a different calc but need to use utility()  during its processing. 
4. These compute  functions are dynamically added to the shader code depending on certain user inputs

E.g. 

group(0) binding(0) var<storage> input_data1:array<f32>; 
group(0) binding(1) var<storage> input_data2:array<f32>;

fn utility(data:array<f32> OR pointer to it) -> f32 {

    // some complex computation running into more than 100 lines of code
    return result;
}

//entry points
fn compute1()  {

    //some processing
   let result1:f32 = utility( input_data1);
   
   ...
   let result2:f32 = utility( input_data2);
   
   //additional processing
}

fn compute2()  {
    //some processing
   let result:f32 = utility( input_data2);
   ...
}

I am not able to find a clean solution to reusing utility() function from different places by passing a dynamically sized array.  Any feedback is appreciated.

Regards,
SK

**Comment by ben-clayton** [comment, marked as answer]
There's currently no way to pass a pointer to data in the `storage` address space. [This will likely become available in a future version of WGSL](https://github.com/gpuweb/gpuweb/issues/3546).
I believe the only options you currently have are to:
1. load the data before calling your `utility` function, and pass the data in as a pointer in the `function` or `private` address-space, then store any writes after the call.
2. or pass an index to `utility`, and switch on this inside the called function to select the storage var you want to read from / write to.

**Reply by skbizport** [reply]
Appreciate suggestions. 

Thx
